File: apis/article.py
from fastapi import APIRouter, Depends, HTTPException, status as fast_status, Query
from core.auth import get_current_user
from core.db import DB
from core.models.base import DATA_STATUS
from core.models.article import Article, ArticleBase
from sqlalchemy import and_, or_, desc
from .base import success_response, error_response
from core.config import cfg
from apis.base import format_search_kw
from core.print import print_warning, print_info, print_error, print_success

router = APIRouter(prefix=f"/articles", tags=["文章管理"])


@router.delete("/clean", summary="清理无效文章(MP_ID不存在于Feeds表中的文章)")
async def clean_orphan_articles(current_user: dict = Depends(get_current_user)):
    session = DB.get_session()
    try:
        from core.models.feed import Feed
        from core.models.article import Article

        # 找出Articles表中mp_id不在Feeds表中的记录
        subquery = session.query(Feed.id).subquery()
        deleted_count = (
            session.query(Article)
            .filter(~Article.mp_id.in_(subquery))
            .delete(synchronize_session=False)
        )

        session.commit()

        return success_response(
            {"message": "清理无效文章成功", "deleted_count": deleted_count}
        )
    except Exception as e:
        session.rollback()
        print(f"清理无效文章错误: {str(e)}")
        raise HTTPException(
            status_code=fast_status.HTTP_201_CREATED,
            detail=error_response(code=50001, message="清理无效文章失败"),
        )


@router.delete("/clean_duplicate_articles", summary="清理重复文章")
async def clean_duplicate(current_user: dict = Depends(get_current_user)):
    try:
        from tools.clean import clean_duplicate_articles

        (msg, deleted_count) = clean_duplicate_articles()
        return success_response({"message": msg, "deleted_count": deleted_count})
    except Exception as e:
        print(f"清理重复文章: {str(e)}")
        raise HTTPException(
            status_code=fast_status.HTTP_201_CREATED,
            detail=error_response(code=50001, message="清理重复文章"),
        )


@router.api_route(
    "",
    summary="获取文章列表",
    methods=["GET", "POST"],
    operation_id="get_articles_list",
)
async def get_articles(
    offset: int = Query(0, ge=0),
    limit: int = Query(5, ge=1, le=100),
    status: str = Query(None),
    search: str = Query(None),
    mp_id: str = Query(None),
    has_content: bool = Query(False),
    current_user: dict = Depends(get_current_user),
):
    session = DB.get_session()
    try:

        # 构建查询条件
        query = session.query(ArticleBase)
        if has_content:
            query = session.query(Article)
        if status:
            query = query.filter(Article.status == status)
        else:
            query = query.filter(Article.status != DATA_STATUS.DELETED)
        if mp_id:
            query = query.filter(Article.mp_id == mp_id)
        if search:
            query = query.filter(format_search_kw(search))

        # 获取总数
        total = query.count()
        # 优先按 publish_at 排序，其次回退到老的 publish_time
        query = (
            query.order_by(Article.publish_at.desc(), Article.publish_time.desc())
            .offset(offset)
            .limit(limit)
        )
        # 分页查询（按发布时间降序）
        articles = query.all()

        # 打印生成的 SQL 语句（包含分页参数）
        print_warning(query.statement.compile(compile_kwargs={"literal_binds": True}))

        # 查询公众号名称
        from core.models.feed import Feed

        mp_names = {}
        for article in articles:
            if article.mp_id and article.mp_id not in mp_names:
                feed = session.query(Feed).filter(Feed.id == article.mp_id).first()
                mp_names[article.mp_id] = feed.mp_name if feed else "未知公众号"

        # 合并公众号名称到文章列表
        article_list = []
        for article in articles:
            article_dict = article.__dict__
            article_dict["mp_name"] = mp_names.get(article.mp_id, "未知公众号")
            article_list.append(article_dict)

        from .base import success_response

        return success_response({"list": article_list, "total": total})
    except HTTPException as e:
        raise e
    except Exception as e:
        raise HTTPException(
            status_code=fast_status.HTTP_406_NOT_ACCEPTABLE,
            detail=error_response(code=50001, message=f"获取文章列表失败: {str(e)}"),
        )


@router.get("/{article_id}", summary="获取文章详情")
async def get_article_detail(
    article_id: str,
    content: bool = False,
    # current_user: dict = Depends(get_current_user)
):
    session = DB.get_session()
    try:
        article = (
            session.query(Article)
            .filter(Article.id == article_id)
            .filter(Article.status != DATA_STATUS.DELETED)
            .first()
        )
        if not article:
            from .base import error_response

            raise HTTPException(
                status_code=fast_status.HTTP_404_NOT_FOUND,
                detail=error_response(code=40401, message="文章不存在"),
            )
        return success_response(article)
    except HTTPException as e:
        raise e
    except Exception as e:
        raise HTTPException(
            status_code=fast_status.HTTP_406_NOT_ACCEPTABLE,
            detail=error_response(code=50001, message=f"获取文章详情失败: {str(e)}"),
        )


@router.delete("/{article_id}", summary="删除文章")
async def delete_article(
    article_id: str, current_user: dict = Depends(get_current_user)
):
    session = DB.get_session()
    try:
        from core.models.article import Article

        # 检查文章是否存在
        article = session.query(Article).filter(Article.id == article_id).first()
        if not article:
            raise HTTPException(
                status_code=fast_status.HTTP_406_NOT_ACCEPTABLE,
                detail=error_response(code=40401, message="文章不存在"),
            )
        # 逻辑删除文章（更新状态为deleted）
        article.status = DATA_STATUS.DELETED
        if cfg.get("article.true_delete", False):
            session.delete(article)
        session.commit()

        return success_response(None, message="文章已标记为删除")
    except Exception as e:
        session.rollback()
        raise HTTPException(
            status_code=fast_status.HTTP_406_NOT_ACCEPTABLE,
            detail=error_response(code=50001, message=f"删除文章失败: {str(e)}"),
        )


@router.get("/{article_id}/next", summary="获取下一篇文章")
async def get_next_article(
    article_id: str, current_user: dict = Depends(get_current_user)
):
    session = DB.get_session()
    try:
        # 获取当前文章的发布时间
        current_article = (
            session.query(Article).filter(Article.id == article_id).first()
        )
        if not current_article:
            raise HTTPException(
                status_code=fast_status.HTTP_404_NOT_FOUND,
                detail=error_response(code=40401, message="当前文章不存在"),
            )

        # 查询发布时间更晚的第一篇文章（优先使用 publish_at）
        if current_article.publish_at:
            next_article = (
                session.query(Article)
                .filter(Article.publish_at > current_article.publish_at)
                .filter(Article.status != DATA_STATUS.DELETED)
                .filter(Article.mp_id == current_article.mp_id)
                .order_by(Article.publish_at.asc())
                .first()
            )
        else:
            # 回退到旧字段（时间戳）
            next_article = (
                session.query(Article)
                .filter(Article.publish_time > current_article.publish_time)
                .filter(Article.status != DATA_STATUS.DELETED)
                .filter(Article.mp_id == current_article.mp_id)
                .order_by(Article.publish_time.asc())
                .first()
            )

        if not next_article:
            raise HTTPException(
                status_code=fast_status.HTTP_406_NOT_ACCEPTABLE,
                detail=error_response(code=40402, message="没有下一篇文章"),
            )

        return success_response(next_article)
    except HTTPException as e:
        raise e
    except Exception as e:
        raise HTTPException(
            status_code=fast_status.HTTP_406_NOT_ACCEPTABLE,
            detail=error_response(code=50001, message=f"获取下一篇文章失败: {str(e)}"),
        )


@router.get("/{article_id}/prev", summary="获取上一篇文章")
async def get_prev_article(
    article_id: str, current_user: dict = Depends(get_current_user)
):
    session = DB.get_session()
    try:
        # 获取当前文章的发布时间
        current_article = (
            session.query(Article).filter(Article.id == article_id).first()
        )
        if not current_article:
            raise HTTPException(
                status_code=fast_status.HTTP_404_NOT_FOUND,
                detail=error_response(code=40401, message="当前文章不存在"),
            )

        # 查询发布时间更早的第一篇文章（优先使用 publish_at）
        if current_article.publish_at:
            prev_article = (
                session.query(Article)
                .filter(Article.publish_at < current_article.publish_at)
                .filter(Article.status != DATA_STATUS.DELETED)
                .filter(Article.mp_id == current_article.mp_id)
                .order_by(Article.publish_at.desc())
                .first()
            )
        else:
            # 回退到旧字段（时间戳）
            prev_article = (
                session.query(Article)
                .filter(Article.publish_time < current_article.publish_time)
                .filter(Article.status != DATA_STATUS.DELETED)
                .filter(Article.mp_id == current_article.mp_id)
                .order_by(Article.publish_time.desc())
                .first()
            )

        if not prev_article:
            raise HTTPException(
                status_code=fast_status.HTTP_406_NOT_ACCEPTABLE,
                detail=error_response(code=40403, message="没有上一篇文章"),
            )

        return success_response(prev_article)
    except HTTPException as e:
        raise e
    except Exception as e:
        raise HTTPException(
            status_code=fast_status.HTTP_406_NOT_ACCEPTABLE,
            detail=error_response(code=50001, message=f"获取上一篇文章失败: {str(e)}"),
        )

File: web_ui/src/api/article.ts
import http from "./http";

/**
 * 文章实体接口
 * @property id 文章ID
 * @property title 文章标题
 * @property content 文章内容
 * @property mp_name 公众号名称
 * @property publish_time 发布时间
 * @property status 文章状态
 * @property link 文章链接
 * @property created_at 创建时间
 */
export interface Article {
    id: number;
    title: string;
    content: string;
    mp_name: string;
    publish_time: string;
    publish_at?: string;
    status: number;
    link: string;
    created_at: string;
}

/**
 * 文章列表查询参数接口
 * @property offset 分页偏移量
 * @property limit 每页数量
 * @property search 搜索关键词
 * @property status 文章状态
 * @property mp_id 公众号ID
 */
export interface ArticleListParams {
    offset?: number;
    limit?: number;
    search?: string;
    status?: number;
    mp_id?: string;
}

/**
 * 文章列表查询结果接口
 * @property code 状态码
 * @property data 文章列表数据
 */
export interface ArticleListResult {
    code: number;
    data: Article[];
}

/**
 * 获取文章列表
 * @param params 查询参数
 * @returns 文章列表结果
 */
export const getArticles = (params: ArticleListParams) => {
    // 转换分页参数
    const apiParams = {
        offset: (params.page || 0) * (params.pageSize || 10),
        limit: params.pageSize || 10,
        search: params.search,
        status: params.status,
        mp_id: params.mp_id,
    };
    return http.get<ArticleListResult>("/wx/articles", {
        params: apiParams,
    });
};

/**
 * 获取文章详情
 * @param id 文章ID
 * @parama 类型 0当前,-1上一篇,1下一篇
 * @returns 文章详情结果
 */
export const getArticleDetail = (id: number, action_type: number) => {
    switch (action_type) {
        case -1:
            return http.get<{ code: number; data: Article }>(
                `/wx/articles/${id}/prev`
            );
        case 1:
            return http.get<{ code: number; data: Article }>(
                `/wx/articles/${id}/next`
            );
        default:
            // 默认获取当前文章详情
            return http.get<{ code: number; data: Article }>(`/wx/articles/${id}`);
            break;
    }
};

/**
 * 获取上一篇文章详情
 * @param id 当前文章ID
 * @returns 上一篇文章详情结果
 */
export const getPrevArticleDetail = (id: number) => {
    return http.get<{ code: number; data: Article }>(`/wx/articles/${id}/prev`);
};

/**
 * 获取下一篇文章详情
 * @param id 当前文章ID
 * @returns 下一篇文章详情结果
 */
export const getNextArticleDetail = (id: number) => {
    return http.get<{ code: number; data: Article }>(`/wx/articles/${id}/next`);
};

/**
 * 删除文章
 * @param id 文章ID
 * @returns 删除结果
 */
export const deleteArticle = (id: number) => {
    return http.delete<{ code: number; message: string }>(`/wx/articles/${id}`);
};

/**
 * 清空所有文章
 * @param id 无实际作用（保留参数）
 * @returns 清空结果
 */
export const ClearArticle = (id: number) => {
    return http.delete<{ code: number; message: string }>(`/wx/articles/clean`);
};

/**
 * 清空重复文章
 * @param id 无实际作用（保留参数）
 * @returns 清空结果
 */
export const ClearDuplicateArticle = (id: number) => {
    return http.delete<{ code: number; message: string }>(
        `/wx/articles/clean_duplicate_articles`
    );
};
